[#17938] xCluster: minor fixes for xCluster architecture page (#18414)
yugabyte/yugabyte-db: docs/content/preview/architecture/docdb-replication/async-replication.md

Although for simplicity, we will describe flows between entire universes, flows are actually composed of streams between pairs of tables, one in each universe, allowing replication of only certain namespaces or tables.

> xCluster is more flexible than a hypothetical scheme whereby read replicas are promoted to full replicas when primary replicas are lost because it does not require the two universes to be tightly coupled. With xCluster, for example, the same table can be split into tablets in different ways in the two universes and (in the future) the two universes will be able to run different versions of YugabyteDB. xCluster also allows for bidirectional replication, which is not possible using read replicas because read replicas cannot take writes.

Reviewer: > the two universes will be able to run different versions of YugabyteDB

That is not true. And we dont know if we will ever support it

Author: Removing that part.

> This mode is an extension of the previous one.  In order to restore consistency, we additionally disallow writes on the target universe and cause reads to read as of a time far enough in the past that all the relevant data from the source universe has already been replicated.

Reviewer: as of a time far enough in the past (typically 250ms)

Reviewer: Its good to have an estimate.

Author: :+1:

> In particular, we pick the time to read as of, _T_, so that all the writes from all the source transactions that will commit at or before time _T_ have been replicated to the target universe.  Put another way, we read as of a time far enough in the past that there cannot be new incoming source writes at or before that time.  This restores consistent reads and ensures source universe transaction results become visible atomically.  Note that we do *not* wait for any current in flight source-universe transactions.

Reviewer: new incoming source writes > new incoming source commits?

Author: :+1:

These are straightforward: when one of these transaction commits, a single Raft log entry is produced containing all of that transaction's writes and its commit time.  This entry in turn is used to generate part of a batch of changes when the poller requests changes.

> Upon receiving the changes, the poller examines each write to see what key it writes to in order to determine which target tablet covers that part of the table then forwards the writes to the appropriate tablets. The commit times of the writes are preserved and the writes are marked as _external_, which prevents them from being further replicated by xCluster, including back in bidirectional cases.

Reviewer 2: ```suggestion
Upon receiving the changes, the poller examines each write to see what key it writes to in order to determine which target tablet covers that part of the table. The poller then forwards the writes to the appropriate tablets. The commit times of the writes are preserved and the writes are marked as _external_, which prevents them from being further replicated, whether to the target or back to the source in bidirectional cases.
```

xCluster safe time is computed for each database by the target-universe master leader as the minimum _xCluster application time_ any tablet in that database has reached.  Pollers determine this time using information from the source tablet servers of the form "once you have fully applied all the changes before this one, your xCluster application time for this tablet will be _T_".

> A source tablet server sends such information when it determines that no active transaction involving that tablet can commit before _T_ and that all transactions involving that tablet that committed before _T_ have application Raft entries that have been previously sent as changes.  It also periodically checks for committed transactions that are missing apply Raft entries and generates such entries for them; this helps xCluster safe time advance faster.

Reviewer: mention every 250ms here too

Author: :+1:

In the future, it may be possible to detect such unsafe constraints and issue a warning, potentially by default.  This is tracked in [#11539](https://github.com/yugabyte/yugabyte-db/issues/11539).

> Note that if you attempt to insert the same row on both universes at the same time to a table that does not have a primary key then you will end up with two rows with the same data.  This is the expected Postgres behavior even with only one universe &mdash; tables without primary keys can have multiple rows with the same data.

Reviewer 2: The standard is to always use the full PostrgreSQL, for whatever reason

Author: will fix...

### Bootstrapping replication

> - Currently, it is your responsibility (see [these instructions](../../../deploy/multi-dc/async-replication/#bootstrap-a-target-universe))   to ensure that a target universe has sufficiently recent updates so   that replication can safely resume.  In the future, bootstrapping the   target universe will be automated, which is tracked in   [#11538](https://github.com/yugabyte/yugabyte-db/issues/11538).

Reviewer 2: ```suggestion
- Currently, it is your responsibility to ensure that a target universe has sufficiently recent updates so that replication can safely resume (for instructions, refer to [Bootstrap a target universe](../../../deploy/multi-dc/async-replication/#bootstrap-a-target-universe)).  In the future, bootstrapping the target universe will be automated, which is tracked in [#11538](https://github.com/yugabyte/yugabyte-db/issues/11538).
```

Author: :+1:

- YSQL geo-partitioning is supported.  Note that you must configure replication on all new partitions manually as DDL changes are not replicated automatically.
- Source and target universes can have different numbers of tablets.
- Tablet splitting is supported on both source and target universes.

> {{< note title="Note" >}}

Reviewer 2: Buried down here. I'd put it up in the second section.
